viabel/optimization.py

```diff
@@ -356,10 +513,379 @@ def optimize(self, n_iters, objective, init_param):
+        log_c = np.mean(fit['log_c'])
+        c = np.exp(log_c)
+        if results:
+            return fit, kappa, c
+        else:
+            return kappa, c
+    
+    def wls(self, x, y, s=9, a=0.25):
+        """
+        weighted least squares
+
+        Parameters
+        ----------
+        x : `numpy_ndarray`
+            Predictor variable
+        y : `numpy_ndarray`
+            Response variable
+        s : `int`, optional
+            Observation weight parameter. The default is 9.
+        a : `int`, optional
+            Power of the weight parameter. The default is 1/4.
+
+        Returns
+        -------
+        b0 : `float`
+            Intercept
+        b1 : `float`
+            Slope
+        """
+        n = y.size
+        x = np.column_stack((np.ones(n),x))
+        w = np.diag(1/(1 + np.arange(n)[::-1]**2/s**2)**a) #weights
+        y = np.reshape(y,(n,1))
+        beta = np.linalg.inv(x.T @ w @ x) @ (x.T @ w @ y)
+        return beta[0], beta[1]
+        
+    def convg_iteration_trend_detection(self, slope):
+        """
+        Detecting the relationship trend between learning
+        rate and number of iterations to reach convergence
+
+        Parameters
+        ----------
+        slope : `float`
+            slope of the fitted regression model
+
+        Returns
+        -------
+        bool
+            Indicating having negative relationship or not
+
+        """
+        if slope < 0:
+            return True
+        else:
+            return False
+        
+
+    def optimize(self, K_max, objective, init_param):
+        """
+        Parameters
+        ----------
+        K_max : `int`
+            Number of iterations of the optimization
+        objective: `function`
+            Function for constructing the objective and gradient function
+        init_param : `numpy.ndarray`, shape(var_param_dim,)
+            Initial values of the variational parameters
+
+        """
+        if not objective.approx.supports_kl:
+            print('WARNING: approximation family does not support KL. Using FASO.',
+                  flush=True)
+            return super().optimize(K_max, objective, init_param)
+        k_new = -1 #Number of iterations at the given learning rate
+        k = 0 #Number of times learning rate decreases
+        k_total = 0 #Total number of iterations
+        k_add = 0 #Iteration number when convergence reached for a fixed step size
+        k_stopped_final = None #Iteration number when stopping rule criteria met
+        sgo = self._sgo
+        diagnostics = self._sgo._diagnostics
+        if isinstance(self._sgo, AveragedRMSProp) or isinstance(self._sgo, AveragedAdam):
+            reg_model = StanModel_cache(model_name='weighted_lin_regression_sgd')
+        else:
+            reg_model = StanModel_cache(model_name='weighted_lin_regression')
+        iterate_average_curr = init_param.copy()
+        iterate_average_curr_hist = [iterate_average_curr]
```

Comment:
a comment or summarizing when / how all of these "history" lists are used would be helpful, especially if they are not used uniformly. Why are some `None` and some `[]` to start?

In fact, would it be possible to use a dictionary from the start to hold all of these histories, since that's how they get returned? You could use a `defaultdict` with default type `list`